Objects/unicodeobject.c

```diff
@@ -15063,106 +15198,263 @@ _PyUnicode_InitTypes(PyInterpreterState *interp)
     return _PyStatus_ERR("Can't initialize unicode types");
 }
 
+static /* non-null */ PyObject*
+intern_static(PyInterpreterState *interp, PyObject *s /* stolen */)
+{
+    // Note that this steals a reference to `s`, but in many cases that
+    // stolen ref is returned, requiring no decref/incref.
+
+    assert(s != NULL);
+    assert(_PyUnicode_CHECK(s));
+    assert(_PyUnicode_STATE(s).statically_allocated);
+    assert(_Py_IsImmortal(s));
+
+    switch (PyUnicode_CHECK_INTERNED(s)) {
+        case SSTATE_NOT_INTERNED:
+            break;
+        case SSTATE_INTERNED_IMMORTAL_STATIC:
+            return s;
+        default:
+            Py_FatalError("_PyUnicode_InternStatic called on wrong string");
+    }
+
+#ifdef Py_DEBUG
+    /* We must not add process-global interned string if there's already a
+     * per-interpreter interned_dict, which might contain duplicates.
+     * Except "short string" singletons: those are special-cased. */
+    PyObject *interned = get_interned_dict(interp);
+    assert(interned == NULL || unicode_is_singleton(s));
+#ifdef Py_GIL_DISABLED
+    // In the free-threaded build, don't allow even the short strings.
+    assert(interned == NULL);
+#endif
+#endif
+
+    /* Look in the global cache first. */
+    PyObject *r = (PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s);
+    /* We should only init each string once */
+    assert(r == NULL);
+    /* but just in case (for the non-debug build), handle this */
+    if (r != NULL && r != s) {
+        assert(_PyUnicode_STATE(r).interned == SSTATE_INTERNED_IMMORTAL_STATIC);
+        assert(_PyUnicode_CHECK(r));
+        Py_DECREF(s);
+        return Py_NewRef(r);
+    }
+
+    if (_Py_hashtable_set(INTERNED_STRINGS, s, s) < -1) {
+        Py_FatalError("failed to intern static string");
+    }
+
+    _PyUnicode_STATE(s).interned = SSTATE_INTERNED_IMMORTAL_STATIC;
+    return s;
+}
```

Comment:
The key nature of the statically allocated strings is that they are tied exclusively to the global runtime state.  They are all interned during runtime init (except for the single-char strings).  Once the runtime is initialized, they are looked up only globally and will never be added to any per-interpreter interned dict.

I thought it would be helpful to split the code here along those lines.  Below (and in the several comments after this) is a rough sketch of what I had in mind.  I'm not saying you need to take that route; what you have seems correct.  Rather, at the least it's a different angle.  Naturally, I'd do all this if it were my PR. 😄

```suggestion
static void
intern_static(PyObject *s)
{
    // The caller is responsible for making sure it's a str object.
    assert(_PyUnicode_STATE(s).statically_allocated);
    assert(_Py_IsImmortal(s));
    assert(PyUnicode_CHECK_INTERNED(s) == SSTATE_NOT_INTERNED);

    /* We should only init each string once */
    assert((PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s) == NULL);

    /* We do not take a ref for the hashtable since we know s is immortal. */
    if (_Py_hashtable_set(INTERNED_STRINGS, s, s) < -1) {
        Py_FatalError("failed to intern static string");
    }

    _PyUnicode_STATE(s).interned = SSTATE_INTERNED_IMMORTAL_STATIC;
}

static /* non-null */ PyObject *
get_interned_static(PyObject *s)
{
    // Note that this steals a reference to `s`, but in many cases that
    // stolen ref is returned, requiring no decref/incref.

    /* The caller is responsible for making sure it's a str object. */
    if (_PyUnicode_STATE(s).statically_allocated) {
        assert(_Py_IsImmortal(s));
#ifndef Py_GIL_DISABLED
        // For now we do intern the single-char strings lazily.
        if (PyUnicode_CHECK_INTERNED(s) == SSTATE_NOT_INTERNED)) {
            assert(PyUnicode_GET_LENGTH(s) == 1
                        && PyUnicode_KIND(s) == PyUnicode_1BYTE_KIND);
            assert(PyUnicode_IS_COMPACT_ASCII(s));
            assert(unicode_is_singleton(s));
            assert(s == LATIN1(*(unsigned char*)PyUnicode_DATA(s)));
            intern_static(s);
            return s;
        }
#endif
        assert(PyUnicode_CHECK_INTERNED(s) == SSTATE_INTERNED_IMMORTAL_STATIC);
        assert((PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s) == s);
        return s;
    }

    PyObject *r = (PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s);
    if (r == NULL) {
        return NULL;
    }
    assert(r != s);  // ...otherwise it would have been handled above.
    assert(_Py_IsImmortal(r));
    assert(_PyUnicode_STATE(r).statically_allocated);
    assert(PyUnicode_CHECK_INTERNED(r) == SSTATE_INTERNED_IMMORTAL_STATIC);
    Py_DECREF(s);
    return Py_NewRef(r);
}

void
_PyUnicode_InternStatic(PyInterpreterState *interp, PyObject *s)
{
    assert(!Py_IsInitialized());  // implies main interpreter
    assert(s != NULL);
    assert(_PyUnicode_CHECK_EXACT(s));

    /* The process-global str objects are all interned before the
     * per-interpreter interned_dict.
     * Except "short string" singletons: those are special-cased. */
    assert(get_interned_dict(interp) == NULL);
#ifdef NDEBUG
    (void)interp;
#endif

    intern_static(s);
}

#ifndef NDEBUG
static int
interned_in_interp(PyInterpreterState *interp, PyObject *s)
{
        // Statically allocated strings must not be in the per-interpreter dict.
        PyObject *interned = get_interned_dict(interp);
        if (interned == NULL || !PyDict_Contains(interned, s)) {
            return 0;
        }
        assert(!PyErr_Occurred());
        return 1;
}
#endif
```

Reply:
Note that `_PyUnicode_InternStatic()` doesn't do the in-place thing since we're guaranteed that `s` will be interned.

Reply:
I'd do that in a follow-up PR, along with interning the single-char strings.

Reply:
Fine with me.

Reply:
See https://github.com/python/cpython/issues/113993#issuecomment-2183246804.